File: src/Vpiska.Application/Event/WebSocket/ChatReceiver.fs
namespace Vpiska.Application.Event.WebSocket

open System.Threading.Tasks
open FSharp.Control.Tasks
open Serilog
open Vpiska.Application.Event
open Vpiska.Domain.Event
open Vpiska.Infrastructure.Websocket

type ChatReceiver(logger: ILogger, storage: UserConnectionsStorage, commandHandler: CommandHandler) =
    
    let queryParamName = "eventId"
    
    let mapErrors errors = errors |> Array.map Errors.mapAppError |> Array.fold (fun acc error -> $"{error}. {acc}") ("")
    
    interface IWebSocketReceiver with
        member this.Receive(connectionId, data, queryParams) =
            task {
                let eventId = queryParams.[queryParamName]
                let mutable context = null
                match storage.TryGetUserInfo (eventId, connectionId, &context) with
                | false -> logger.Warning("can't get userInfo while receive message. EventId: {eventId}, ConnectionId: {connectionId}", eventId, connectionId)
                | true ->
                    let command = { EventId = eventId; UserId = context.UserId; Message = data |> WebSocketSerializer.deserializeToString }
                                  |> Command.SendChatMessage
                    match! commandHandler.Handle command with
                    | Error errors ->
                        let message = mapErrors errors
                        logger.Warning("can't send message. Message: {message}", message)
                    | Ok _ -> return ()
            } :> Task

File: src/Vpiska.Infrastructure/WebSocket/ChatReceiver.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Vpiska.Domain.Event.Commands.ChatMessageCommand;
using Vpiska.Domain.Event.Models;
using Vpiska.Domain.Interfaces;
using Vpiska.WebSocket;

namespace Vpiska.Infrastructure.WebSocket
{
    internal sealed class ChatReceiver : IWebSocketReceiver
    {
        private readonly ILogger<ChatReceiver> _logger;
        private readonly ICommandHandler<ChatMessageCommand> _commandHandler;

        public ChatReceiver(ILogger<ChatReceiver> logger, ICommandHandler<ChatMessageCommand> commandHandler)
        {
            _logger = logger;
            _commandHandler = commandHandler;
        }

        public Task Receive(Guid connectionId, string route, string message,
            Dictionary<string, string> identityParams,
            Dictionary<string, string> queryParams)
        {
            switch (route)
            {
                case "chatMessage":
                {
                    var command = new ChatMessageCommand()
                    {
                        EventId = queryParams["eventId"],
                        ChatMessage = new ChatMessage()
                        {
                            UserId = identityParams["Id"],
                            UserName = identityParams["Name"],
                            UserImageId = identityParams["ImageId"],
                            Message = message
                        }
                    };
                    return _commandHandler.HandleAsync(command);
                }
                default:
                    _logger.LogWarning("Unknown receiver route {}", route);
                    return Task.CompletedTask;
            }
        }
    }
}